pennylane/interfaces/jacobian_products.py

```diff
@@ -215,3 +229,318 @@ def compute_jacobian(self, tapes: Batch):
         )
         results = self._inner_execute(jac_tapes)
         return tuple(batch_post_processing(results))
+
+
+class DeviceDerivatives(JacobianProductCalculator):
+    """Calculate jacobian products via a device provided jacobian.  This class relies on either ``qml.Device.gradients`` or
+    ``qml.devices.Device.compute_derivatives``.
+
+    Args:
+
+        device (Union[pennylane.Device, pennylane.devices.Device]): the device for execution and derivatives.
+            Must support first order gradients with the requested configuration.
+        execution_config (pennylane.devices.ExecutionConfig): a datastructure containing the options needed to fully
+           describe the execution. Only used with :class:`pennylane.devices.Device` from the new device interface.
+        gradient_kwargs (dict): a dictionary of keyword arguments for the gradients. Only used with a :class:`~.pennylane.Device`
+            from the old device interface.
+
+    **Examples:**
+
+    >>> device = qml.device('default.qubit')
+    >>> config = qml.devices.ExecutionConfig(gradient_method="adjoint")
+    >>> jpc = DeviceDerivatives(device, config, {})
+
+    This same class can also be used with the old device interface.
+
+    >>> device = qml.device('lightning.qubit', wires=5)
+    >>> gradient_kwargs = {"method": "adjoint_jacobian"}
+    >>> jpc_lightning = DeviceDerivatives(device, gradient_kwargs=gradient_kwargs)
+
+    **Technical comments on caching and calculating the gradients on execution:**
+
+    In order to store results and Jacobians for the backward pass during the forward pass,
+    the ``_jacs_cache`` and ``_results_cache`` properties are ``LRUCache`` objects with a maximum size of 10.
+    In the current execution pipeline, only one batch will be used per instance, but a size of 10 adds some extra
+    flexibility for future uses.
+
+    Note that since the hash and equality of :class:`~.QuantumScript` is based on object location, not contents. So batches
+    with identically looking :class:`~.QuantumScript` s that are different instances will be cached separately.
+
+    When a forward pass with :meth:`~.execute` is called, both the results and the jacobian for the object are stored.
+
+    >>> tape = qml.tape.QuantumScript([qml.RX(1.0, wires=0)], [qml.expval(qml.PauliZ(0))])
+    >>> batch = (tape, )
+    >>> with device.tracker:
+    ...     results = jpc.execute(batch )
+    >>> results
+    (0.5403023058681398,)
+    >>> device.tracker.totals
+    {'execute_and_derivative_batches': 1, 'executions': 1, 'derivatives': 1}
+    >>> jpc._jacs_cache
+    LRUCache({5660934048: (array(-0.84147098),)}, maxsize=10, currsize=1)
+
+    Then when the vjp, jvp, or jacobian is requested, that cached value is used instead of requesting from
+    the device again.
+
+    >>> with device.tracker:
+    ...     vjp = jpc.compute_vjp(batch , (0.5, ) )
+    >>> vjp
+    (array([-0.42073549]),)
+    >>> device.tracker.totals
+    {}
```

Comment:
@albi3ro for a user working with a QNode, where `diff_method="adjoint"`, will the cache be reset between each execution?

Or will requesting the same derivatives across multiple executions result in the cache being retained and used?

I believe the current implementation, the cache is contained within the execution 'closure', so is only active per individual (forward+backward) pass

Reply:
A new instance is created on each call to `qml.execute`.

Reply:
I could potentially add a `cache` keyword argument and allow it to be reused between computations.

Reply:
> A new instance is created on each call to qml.execute.

Got it! 

> I could potentially add a cache keyword argument and allow it to be reused between computations.

one to double check with @trbromley beforehand, because it could be that there is no user need for this right now :) I was just more curious if this was a change in existing behaviour or not.

Reply:
Basically this is the way to save jacobians for the backward pass.

Reply:
My preference is to preserve existing behaviour (as closely as possible), as I don't recall hearing anything from users asking for changes.